### ISSUE TITLE
x = x[xc[xi]]

### ISSUE THREAD

**Issue by carlsummer** [ultralytics/yolov5#2729]
for xi, x in enumerate(prediction):  # image index, image inference
        # Apply constraints
        # x[((x[..., 2:4] < min_wh) | (x[..., 2:4] > max_wh)).any(1), 4] = 0  # width-height
        x = x[xc[xi]]  # confidence

        # Cat apriori labels if autolabelling
        if labels and len(labels[xi]):
            l = labels[xi]
            v = torch.zeros((len(l), nc + 5), device=x.device)
            v[:, :4] = l[:, 1:5]  # box
            v[:, 4] = 1.0  # conf
            v[range(len(l)), l[:, 0].long() + 5] = 1.0  # cls
            x = torch.cat((x, v), 0)

        # If none remain process next image
        if not x.shape[0]:
            continue

**Comment by carlsummer** [comment]
x = x[xc[xi]] is very slow

**Comment by glenn-jocher** [comment]
@carlsummer hi thanks for the feedback! This line is in our NMS function. Do you have profiling results indicating a speed problem with this line?

NMS time (including this line) is generally <1 millisecond per image, and if I attach a line profiler here when running detect.py I don't see any issue here (0.16 ms on L443 for 2 images). 
<img width="1764" alt="Screenshot 2021-04-07 at 18 36 31" src="https://user-images.githubusercontent.com/26833433/113902536-487bd980-97d0-11eb-9a3a-c0c1b68e4c76.png">

If you have a faster alternative implementation though please submit a PR so everyone can benefit. Thank you!

**Comment by carlsummer** [comment]
![image](https://user-images.githubusercontent.com/16281965/114116334-14f79700-9917-11eb-8513-71037b86e515.png)

**Comment by carlsummer** [comment]
@glenn-jocher

**Comment by HonestyBrave** [comment]
hi, @glenn-jocher this is my colleague feedback, we use the same code and same picture(the picture num is a total of 144), but the diff is the device platform:
Tesla V100-PCIE： 4.555342435836792 s
Tesla T4：9.126741409301758 s

because we use picture num is 144, so, is not every picture need much time, we print the time, as below:
![image](https://user-images.githubusercontent.com/16200506/113954249-16a85880-984c-11eb-8c3d-a636d5a70a0b.png)

**Comment by glenn-jocher** [comment]
@HonestyBrave thanks for the screenshots! Yes 1.8 seconds is a lot of time for this line, I'm pretty surprised. Can you share the command you used to produce these results?

The profiling results I shared before were from `python detect.py`. I'll do a more extensive test here, I'll re-run the same experiment for 3 commands (master branch no changes to code) to see if I can spot the slowdown. Note this is on CPU. The results are consistent, they should < 0.2ms per image for this line.

One thing that may be causing this could be improper timestamps on CUDA tensors. When you profile on GPU, you need to use torch.cuda.synchronize() before each time.time(). We have a helper function that provides CUDA-accurate times you may want to use instead of time.time():
https://github.com/ultralytics/yolov5/blob/fca5e2a48fb526b57bda0c66be6b7ac1aaa8d83d/utils/torch_utils.py#L89-L94


## python detect.py
<img width="1000" alt="Screenshot 2021-04-09 at 13 21 29" src="https://user-images.githubusercontent.com/26833433/114172950-a7f3f980-9936-11eb-8352-8b9f856bad80.png">

## python hubconf.py
<img width="1000" alt="Screenshot 2021-04-09 at 13 22 02" src="https://user-images.githubusercontent.com/26833433/114172959-aaeeea00-9936-11eb-9db1-2c11ab27fbf1.png">

## python test.py --data coco128.yaml
Note had to set dataloader workers=0 for lineprofiler to work.
<img width="1764" alt="Screenshot 2021-04-09 at 13 31 13" src="https://user-images.githubusercontent.com/26833433/114173705-df16da80-9937-11eb-974f-b7953507a3df.png">

**Comment by glenn-jocher** [comment]
@HonestyBrave to test on GPU I used a Colab P100 instance and used time_synchronized. Each call is about 0.1 ms, so everything seems fine I think.

<img width="1500" alt="Screenshot 2021-04-09 at 13 39 24" src="https://user-images.githubusercontent.com/26833433/114174653-34072080-9939-11eb-9d0e-3a55e6df4cc8.png">

**Comment by yakaib** [comment]
I meet the same problem, when the code runs on a high performance computer,  its ok, but when the code runs on a low performance computer(nvidia nano), this problem occures.

**Comment by yakaib** [comment]
a very large batch size may also cause this problem

**Comment by glenn-jocher** [comment]
@yakaib Thanks for the information! Running on a lower performance computer with a large batch size could indeed be a contributing factor to this issue. The computation and memory requirements increase with larger batch sizes, which can lead to decreased performance on less powerful hardware.

You may want to consider reducing the batch size when running on lower performance hardware to see if it improves the inference speed. Additionally, optimizing the model architecture or using a lighter weight model could also help improve performance on less powerful devices.

Let me know if you need further assistance!
